engine/hooks/hooks_router.go

```diff
@@ -48,6 +57,71 @@ func (s *Service) initRouter(ctx context.Context) {
 	r.Handle("/task/{uuid}/execution/{timestamp}/stop", nil, r.POST(s.postStopTaskExecutionHandler))
 }
 
+func (s *Service) CheckHeaderToken(headerName string) service.Middleware {
+	return func(ctx context.Context, w http.ResponseWriter, req *http.Request, rc *service.HandlerConfig) (context.Context, error) {
+		tokenHeaderValue := req.Header.Get(headerName)
+		if tokenHeaderValue == "" {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to check token")
+		}
+		vars := mux.Vars(req)
+		uuid := vars["uuid"]
+
+		hook, err := s.Client.RepositoryHook(ctx, uuid)
+		if err != nil {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to retrieve sign key")
+		}
+
+		defer req.Body.Close()
+
+		if tokenHeaderValue != hook.HookSignKey {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "token mismatch")
+		}
+
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to check signature")
+		}
+
+		ctx = context.WithValue(ctx, "body", body)
+		return ctx, nil
+	}
+}
+
+func (s *Service) CheckHmac256Signature(headerName string) service.Middleware {
+	return func(ctx context.Context, w http.ResponseWriter, req *http.Request, rc *service.HandlerConfig) (context.Context, error) {
+		signHeaderValue := req.Header.Get(headerName)
+		if signHeaderValue == "" {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to check signature")
+		}
+		vars := mux.Vars(req)
+		uuid := vars["uuid"]
+
+		hook, err := s.Client.RepositoryHook(ctx, uuid)
+		if err != nil {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to retrieve sign key")
+		}
+
+		defer req.Body.Close()
+		body, err := io.ReadAll(req.Body)
+		if err != nil {
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "unable to check signature")
+		}
+
+		// Create a new HMAC by defining the hash type and the key (as byte array)
+		h := hmac.New(sha256.New, []byte(hook.HookSignKey))
+		h.Write(body)
+		sha := hex.EncodeToString(h.Sum(nil))
+
+		if strings.TrimPrefix(signHeaderValue, "sha256=") != sha {
+			log.Error(ctx, "signature mismatch: got %s, compute %s", signHeaderValue, sha)
+			return ctx, sdk.NewErrorFrom(sdk.ErrUnauthorized, "wrong signature")
+		}
+
+		ctx = context.WithValue(ctx, "body", body)
```

Comment:
don't put the body in the context, you have to reset a readable body in the request itself